Update scopes docs (#7071)
auth0/docs: articles/scopes/current/index.md

---
title: Scopes
description: Understand the principle of scopes and explore general examples of their use.
topics:
  - scopes
contentType:
  - concept
  - index
useCase:
  - development
  - add-login
  - secure-api
  - call-api
---
# Scopes

> Scopes let you control the type of access your users need; they are a method of limiting the access given to a user by a [token](/tokens). 

Reviewer: I'm not sure I'd say "by a token" here.

`Scopes are a method by which you can control the specific types of access to resources that are granted to the application on the user's behalf.` or if you want to mention tokens, `Scopes are a method by which you can control the specific types of access a user's token will grant.` or something maybe? The token isn't the thing causing the limitation (although it is the tool by which they gain access).

When an app requests access to a resource through an authorization server, it uses the `scope` parameter to specify what type of access it desires, and the authorization server uses the `scope` parameter to respond with the type of access that was actually granted (if the granted access level was different from what was requested).

Generally, you use scopes in two ways:

> * In an [application](/applications), to verify the identity of a user and get basic profile information about the user, such as their email or picture. In this scenario, the scopes available to you include those implemented by the [OpenID Connect](/protocols/oidc) protocol. For details, refer to [OpenID Connect Scopes](/scopes/current/oidc-scopes).

Reviewer: So, the slight wording change at the start makes it sound like this use case occurs within an application. You're using OIDC, but you're still querying an API endpoint from the application, and then getting back standard claims with user info.

* In an [API](/apis), to implement granular access control. In this case, you need to define custom scopes for your API and then identify these scopes so that calling applications can use them. For details, refer to [API Scopes](/scopes/current/api-scopes).

## Best practices

Understand your use case and choose the most restrictive scope possible. Are you establishing a user's identity or asking the user to allow you to interact with their data? There's a big difference between importing a user's Facebook profile information and posting to their wall. Only request what you absolutely need. By doing so, you are also more likely to gain user consent since users are more likely to grant access for limited, clearly-specified scopes.

> 

Reviewer: Should be single returns between sections

## Requested scopes versus granted scopes

> Remember that a user gets to consent to the access level you are requesting. While usually the scopes returned will be identical to the scopes you requested, users can edit their scopes (both during and after initial consent), thereby granting your app less access than you requested. 

Reviewer: This is a question, not a critique - but is this true? `...users can edit their scopes (both during and after initial consent)...`

I would assume that this would depend entirely on the application you had built? Surely in some cases they just have to approve (or deny) the application's request to access (whatever scopes)? WDYT?

Author: I can actually authenticate using google, authorizing an app for certain scopes and then go back into Google and remove some of those scopes. I can do the same in Githhub. This is what I was thinking about--the fact that at any time, I could go back and be like, "NO, app, you are no longer allowed to look at my repos." So it's not really dependent on YOUR application. But it is dependent on the resource you're requesting access to. I will rephrase to make that clearer.

---
title: Permissions
description: Understand the principle of permissions (scopes) and explore general examples of their use.
topics:
  - scopes
  - permissions
contentType:
  - concept
  - index
useCase:
  - development
  - add-login
  - secure-api
  - call-api
---

> # Permissions

Reviewer 2: I'm somewhat troubled by the almost blanket replacement of `scope` with `permission` in this doc page. I feel like we're conflating two separate concepts (see [Vittorio's blog post](https://auth0.com/blog/on-the-nature-of-oauth2-scopes/) for more details). 

Permissions in this conversation are really _user permissions_. They answer the question: 
_What can this user do with resource X?_

Whereas scopes are more of an OAuth construct designed to address _delegated access_. They attempt to answer the question: 
_What has the user allowed this application to do on its behalf with resource X?_

I feel like we really need two separate docs pages to address these concepts separately; otherwise, we're inviting a bit of a confusion mess.

I realize that we're potentially going to release an option with the new AuthZ feature where a customer could treat scopes as if they were permissions - a sort of short cut for the purposes of simplification. But I don't think that means we merge the concepts in the docs. We still need separate discussions and we need to provide proper guidance as to which option a customer should choose when modeling their API authorization.

Author: Reverted the terminology for this release as further review and convos have revealed a need to more thoroughly define when to use "scopes" and when to use "permissions". Spoke with Steve and agreed that terminology changes would be held for release with additional Authz docs changes.

Different pieces of user information are often stored across a number of online resources. Users may upload and store photos with a service like Flickr, keep digital files on Dropbox, and store contacts and events in Google Calendar or on Facebook.

> Often, new applications will want to make use of the information that has already been created in an online resource. To do so, the application must ask for authorization to access this information on a user's behalf. Permissions (also known as _scopes_ in the OAuth 2.0 specs) define the specific actions third-party applications can be allowed to do on a user's behalf.

Reviewer 3: `in the OAuth 2.0 specs`

Maybe link to the spec?

Author: Linked.

## Ways to use permissions

> When an app requests permission to access a resource through an authorization server, it uses the `scope` parameter to specify what type of permission it needs, and the authorization server uses the `scope` parameter to respond with the type of permission that was actually granted (if the granted permission was different from what was requested).

Reviewer 2: The phrase "type of permission" here is a bit confusing. I think it should just be "permissions" as "type of" infers there's some higher level generalization, but there's not. For example, a type of permission might be _permissions for reading_ and another might be _permissions for writing_. But the `scope` parameter can contain any mix of "types" of permissions (eg. `scope=read:foo write:foo`).

```suggestion
When an app requests permission to access a resource through an authorization server, it uses the `scope` parameter to specify which permissions it needs, and the authorization server uses the `scope` parameter to respond with the permissions that were actually granted (if the granted permission was different from what was requested).
```

> When an app requests permission to access a resource through an authorization server, it uses the `scope` parameter to specify what type of permission it needs, and the authorization server uses the `scope` parameter to respond with the type of permission that was actually granted (if the granted permission was different from what was requested).

Reviewer 3: > and the authorization server uses the `scope` parameter

This is true, but in our case we might actually use different/additional claims. We'll have to update the doc if we change this during beta.

Author: Agreed. This will be a change we make during doc updates specifically related to the Authz release for when it comes out of beta. This doc will go out prior to that, so we can't make those changes just yet.

Generally, you use permissions in three ways:

* From an [application](/applications), to verify the identity of a user and get basic profile information about the user, such as their email or picture. In this scenario, the permissions available to you include those implemented by the [OpenID Connect](/protocols/oidc) protocol. For details, refer to [OpenID Connect Permissions](/scopes/current/oidc-scopes).

> * In an [API](/apis), to implement granular access control. In this case, you need to define custom permissions for your API and then identify these permissions so that calling applications can use them. For details, refer to [API Permissions](/scopes/current/api-scopes).

Reviewer 3: I would omit `granular`, since it is unclear and up to the API implementer how granular permissions should be (and there are appropriate levels of granularity in fact - background: https://auth0.com/blog/on-the-nature-of-oauth2-scopes/)

Author: Removed.

* From an application, to call an API that has implemented its own custom permissions. In this case, you need to know which custom permissions are defined for the API you are calling. For an example of calling a custom API from an application, see [Sample Use Cases: Permissions and Claims](/scopes/current/sample-use-cases#request-custom-API-access)

## Best practices

Understand your use case and choose the most restrictive permissions possible. 

If you are requesting permission, make sure you ask for enough access for your application to function, but only request what you absolutely need. Are you establishing a user's identity or asking the user to allow you to interact with their data? There's a big difference between importing a user's Facebook profile information and posting to their wall. By only requesting what you need, you are more likely to gain user consent since users are more likely to grant access for limited, clearly-specified permissions. 

Similarly, when creating custom permissions for an API, consider what levels of granular access applications may need and design accordingly.

> ## Requested scopes versus granted scopes

Reviewer 3: I feel like the heading should reflect the `Permission` term.

Author: Fixed.

> Remember that a user gets to consent to the permissions being requested. While usually the permissions returned will be identical to those requested, users can edit granted permissions (both during initial consent and sometimes after, depending on the resource), thereby granting an app less access than it requested. 

Reviewer 2: >Remember that a user gets to consent to the permissions being requested

This is really only the case with 3rd-party clients. If a 1st-party client is performing a call to `/authorize`, the user is never prompted (with an OAuth Consent page) to approve all the scopes being requested by the client. We should probably make this distinction here.

As an application developer, you should be aware of this possibility and handle these cases in your app. For example, your app could warn the user that they will see reduced functionality. It could also send the user back through the authorization flow to ask for additional permissions. But again, remember that users can always say no.

## Keep reading

> - [OpenID Connect Permissions](/scopes/current/oidc-scopes)

Reviewer 3: IMHO, I don't think we should call these `OpenID Connect Permissions` because the `scopes` terminology is embedded in the OIDC spec itself, so OpenID Connect Scopes seems to make sense in this context. But maybe for consistency we should use the same term, I don't know just expressing my 2 cents.

Note this comment applies throughout all the pages, so I won't repeat it elsewhere.

Author: Reverted the terminology for this release as further review and convos have revealed a need to more thoroughly define when to use "scopes" and when to use "permissions". Spoke with Steve and agreed that terminology changes would be held for release with additional Authz docs changes.
